### PR TITLE
Use proposal filtering

### PR DESCRIPTION
**Pull request by JunAishima**
Use the API endpoint v1/proposals which can have added parameters, which results in an increase in query speed (tests with the 2024-2 cycle and several beamlines) indicate going from 200+ sec to get all of the proposals for the cycle and beamlines combination to less than 1 sec.

### CODE REVIEW

File: nsls2api_lib.py [hunk 1]
```diff
@@ -24,6 +24,35 @@ def get_all_cycles():
...
     return get_from_api(f"facility/nsls2/cycle/{cycle}/proposals")['proposals']
+
+# return true if any of check_instruments are in proposal_instruments
+def check_instruments_in_proposal(proposal_instruments, check_instruments):
+    for instrument in check_instruments:
+        if instrument.upper() in proposal_instruments:
+            return True
+    return False
+
+def get_proposals_for_cycle_instruments(cycle, instruments):
```
Comment (vshekar) [inline nsls2api_lib.py:35]:
```suggestion
def get_proposal_ids_for_cycle_instruments(cycle, instruments):
```
Maybe best to specify the function returns only IDs

File: nsls2api_lib.py [hunk 2]
```diff
@@ -24,6 +24,35 @@ def get_all_cycles():
...
+    instruments_string = ""
+    for index, instrument in enumerate(instruments):
+        if index > 0:
+            instruments_string += "&"
+        instruments_string += f"beamline={instrument}"
```
Comment (vshekar) [inline nsls2api_lib.py:40]:
```suggestion
    instruments_string = "beamline=" + '&beamline='.join(instruments)
```
The above should do the same thing

File: nsls2api_lib.py [hunk 3]
```diff
@@ -24,6 +24,35 @@ def get_all_cycles():
...
+    cycle_string = f"cycle={cycle}"
+    proposal_query = f"proposals/?{instruments_string}&{cycle_string}&facility=nsls2"
```
Comment (vshekar) [inline nsls2api_lib.py:42]:
```suggestion
    proposal_query = f"proposals/?{instruments_string}&cycle={cycle}&facility=nsls2"
```

File: nsls2api_lib.py [hunk 4]
```diff
@@ -24,6 +24,35 @@ def get_all_cycles():
...
+    items_per_page = 50
+    page = 1
+    while 1:
+        proposals = get_from_api(f"proposals/?{instruments_string}&{cycle_string}&facility=nsls2&page_size={items_per_page}&page={page}")
+        for proposal in proposals['proposals']:
+            proposals_to_return.add(proposal['proposal_id'])
+        if proposals['count'] == items_per_page:
+            page += 1
+        else:
+            break
```
Comment (vshekar) [inline nsls2api_lib.py:53]:
```suggestion
        if proposals['count'] != items_per_page:
            break
        page += 1
```

File: ispyb-update-all-proposals [hunk 5]
```diff
@@ -28,42 +31,26 @@ def main(args=[]):
...
-    print(f"Starting with a total of {len(all_proposals)} proposals")
-    ispyb_proposals = set()
-    for proposal in sorted(all_proposals):
-        proposal_info = nsls2api_lib.get_proposal_info(proposal)
-        if "error_message" in proposal_info.keys():
-            continue  # no useful proposal info - such as a missing one
-        if nsls2api_lib.is_ispyb_instrument(proposal_info["instruments"]):
-            print(f"adding {proposal_info['proposal_id']}")
-            ispyb_proposals.add(proposal_info["proposal_id"])
+    ispyb_proposals = nsls2api_lib.get_all_proposal_ids_for_instruments(
```
Comment (RobertSchaffer1) [inline ispyb-update-all-proposals:36]: I don't see this function in the https://github.com/NSLS2/nsls2api-client-for-mx/pull/2/files pr, am I looking in the wrong spot?

File: ispyb_lib.py [hunk 6]
```diff
@@ -4,26 +4,30 @@
...
-  cursor.execute(q)
-  try:
-    return list(cursor.fetchall())
-  except TypeError:
-    return 0
+    cursor.execute(q)
+    try:
+        return list(cursor.fetchall())
+    except TypeError:
+        return 0
```
Comment (vshekar) [inline ispyb_lib.py:21]: Curious under what circumstances is a TypeError raised? From my reading fetchall() should always return an iterable so should be convertable to an empty list

**Review by RobertSchaffer1 — APPROVED** [review]
All looks okay to me, assuming Shekar is also satisfied.

**Review by vshekar — APPROVED** [review]

**Pull request merged**